### PR TITLE
Add `DbtCloudJobRunAsyncSensor`

### PR DESCRIPTION
**Pull request by phanikumv**
This PR donates `DbtCloudJobRunAsyncSensor` from [astronomer-providers ](https://github.com/astronomer/astronomer-providers)repo to OSS Airflow

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/providers/dbt/cloud/sensors/dbt.py [hunk 1]
```diff
@@ -64,3 +72,56 @@ def poke(self, context: Context) -> bool:
...
 
         return job_run_status == DbtCloudJobRunStatus.SUCCESS.value
+
+
+class DbtCloudJobRunAsyncSensor(DbtCloudJobRunSensor):
+    """
+    Checks the status of a dbt Cloud job run asynchronously.
+
+    .. seealso::
+        For more information on sync Sensor DbtCloudJobRunAsyncSensor, take a look at the guide::
```
Comment (josh-fell) [inline airflow/providers/dbt/cloud/sensors/dbt.py]:
```suggestion
        For more information on the DbtCloudJobRunAsyncSensor, take a look at the guide:
```

File: airflow/providers/dbt/cloud/sensors/dbt.py [hunk 2]
```diff
@@ -64,3 +72,56 @@ def poke(self, context: Context) -> bool:
...
+    :param dbt_cloud_conn_id: The connection identifier for connecting to dbt Cloud.
+    :param run_id: The job run identifier.
+    :param account_id: The dbt Cloud account identifier.
+    :param timeout: Time in seconds to wait for a job run to reach a terminal status. Defaults to 7 days.
+    """
+
+    def __init__(
+        self,
+        *,
+        poll_interval: float = 5,
```
Comment (josh-fell) [inline airflow/providers/dbt/cloud/sensors/dbt.py:90]: Would mind adding this param to the docstring as well?

File: airflow/providers/dbt/cloud/sensors/dbt.py [hunk 3]
```diff
@@ -64,3 +72,56 @@ def poke(self, context: Context) -> bool:
...
+                run_id=self.run_id,
+                conn_id=self.dbt_cloud_conn_id,
+                account_id=self.account_id,
+                poll_interval=self.poll_interval,
+                end_time=end_time,
+            ),
+            method_name="execute_complete",
+        )
+
+    def execute_complete(self, context: Context, event: Dict[str, Any]) -> int:
```
Comment (josh-fell) [inline airflow/providers/dbt/cloud/sensors/dbt.py]:
```suggestion
    def execute_complete(self, context: Context, event: dict[str, Any]) -> int:
```
PEP 563 should handle this.

File: airflow/providers/dbt/cloud/sensors/dbt.py [hunk 4]
```diff
@@ -64,3 +72,56 @@ def poke(self, context: Context) -> bool:
...
+        timeout: float = 60 * 60 * 24 * 7,
+        **kwargs: Any,
+    ):
+        self.poll_interval = poll_interval
+        self.timeout = timeout
+        super().__init__(**kwargs)
+
+    def execute(self, context: Context) -> None:
+        """Defers to Trigger class to poll for state of the job run until
+        it reaches a failure state or success state"""
```
Comment (josh-fell) [inline airflow/providers/dbt/cloud/sensors/dbt.py]:
```suggestion
        """
        Defers to Trigger class to poll for state of the job run until
        it reaches a failure state or success state.
        """
```

**Review by josh-fell — COMMENTED** [review]
It would be great to add unit tests for this as well.

File: airflow/providers/dbt/cloud/sensors/dbt.py [hunk 5]
```diff
@@ -64,3 +72,56 @@ def poke(self, context: Context) -> bool:
...
+        )
+
+    def execute_complete(self, context: Context, event: Dict[str, Any]) -> int:
+        """
+        Callback for when the trigger fires - returns immediately.
+        Relies on trigger to throw an exception, otherwise it assumes execution was
+        successful.
+        """
+        if event["status"] in ["error", "cancelled"]:
+            raise AirflowException(event["message"])
```
Comment (dimberman) [inline airflow/providers/dbt/cloud/sensors/dbt.py]: Should we wrap this error? Something like "Error in dbt: {event[message]}"

That way users know that the error is not based in Airflow and is instead in dbt.

**Review by dimberman — APPROVED** [review]
A few minor comments but overall this looks good to me 👍

File: airflow/providers/dbt/cloud/sensors/dbt.py (hunk 1 above)
Comment (phanikumv) [reply airflow/providers/dbt/cloud/sensors/dbt.py]: fixed

File: airflow/providers/dbt/cloud/sensors/dbt.py (hunk 2 above)
Comment (phanikumv) [reply airflow/providers/dbt/cloud/sensors/dbt.py:90]: fixed

File: airflow/providers/dbt/cloud/sensors/dbt.py (hunk 4 above)
Comment (phanikumv) [reply airflow/providers/dbt/cloud/sensors/dbt.py]: fixed

File: airflow/providers/dbt/cloud/sensors/dbt.py (hunk 3 above)
Comment (phanikumv) [reply airflow/providers/dbt/cloud/sensors/dbt.py]: fixed

File: airflow/providers/dbt/cloud/sensors/dbt.py (hunk 5 above)
Comment (phanikumv) [reply airflow/providers/dbt/cloud/sensors/dbt.py]: fixed

**Comment by phanikumv** [comment]
> It would be great to add unit tests for this as well.

added unit tests

**Review by josh-fell — APPROVED** [review]

**Pull request merged**